rubix-spi/src/main/java/com/qubole/rubix/spi/fop/ObjectPoolPartition.java

```diff
@@ -50,115 +53,83 @@ public ObjectPoolPartition(ObjectPool<T> pool, PoolConfig config,
+    try {
+      for (int i = 0; i < config.getMinSize(); i++) {
+          T object = objectFactory.create(host, socketTimeout, connectTimeout);
+          objectQueue.add(new Poolable<>(object, pool, host));
+          totalCount.incrementAndGet();
       }
     }
+    catch (Exception e) {
+      // skipping logging the exception as factories are already logging.
+    }
   }
 
   public void returnObject(Poolable<T> object)
   {
-    if (!objectFactory.validate(object.getObject())) {
-      log.debug(String.format("Invalid object...removing: %s ", object));
-      decreaseObject(object);
-      // Compensate for the removed object. Needed to prevent endless wait when in parallel a borrowObject is called
-      increaseObjects(1, false);
-      return;
-    }
+    try {
+      if (!objectFactory.validate(object.getObject())) {
+        log.debug(String.format("Invalid object...removing: %s ", object));
+        decreaseObject(object);
+        return;
+      }
 
-    log.debug(String.format("Returning object: %s to queue. Queue size: %d", object, objectQueue.size()));
-    if (!objectQueue.offer(object)) {
-      log.warn("Created more objects than configured. Created=" + totalCount + " QueueSize=" + objectQueue.size());
-      decreaseObject(object);
+      log.debug(String.format("Returning object: %s to queue. Queue size: %d", object, objectQueue.size()));
+      if (!objectQueue.offer(object)) {
+        log.warn("Created more objects than configured. Created=" + totalCount + " QueueSize=" + objectQueue.size());
+        decreaseObject(object);
+      }
+    }
+    finally {
+      takeSemaphore.release();
     }
   }
 
   public Poolable<T> getObject(boolean blocking)
   {
-    if (objectQueue.size() == 0) {
-      // increase objects and return one, it will return null if pool reaches max size or if object creation fails
-      Poolable<T> object = increaseObjects(this.config.getDelta(), true);
-
-      if (object != null) {
-        return object;
-      }
-
-      if (totalCount == 0) {
-        // Could not create objects, this is mostly due to connection timeouts hence no point blocking as there is not other producer of sockets
-        throw new RuntimeException("Could not add connections to pool");
-      }
-      // else wait for a connection to get free
-    }
-
-    Poolable<T> freeObject;
+    Poolable<T> object;
     try {
       if (blocking) {
-        freeObject = objectQueue.take();
+        takeSemaphore.acquire();
+        object = objectQueue.take();
       }
       else {
-        freeObject = objectQueue.poll(config.getMaxWaitMilliseconds(), TimeUnit.MILLISECONDS);
-        if (freeObject == null) {
-          throw new RuntimeException("Cannot get a free object from the pool");
-        }
+        takeSemaphore.tryAcquire(config.getMaxWaitMilliseconds(), TimeUnit.MILLISECONDS);
+        object = tryGetObject();
       }
+      object.setLastAccessTs(System.currentTimeMillis());
```

Comment:
why is this set while giving the object? what if this object is being used for more than the max-idle time?
shouldn't this be set to some sentinel value while giving the object, and to a timestamp while release of the object of the pool?

Reply:
We should fix this separately, I will open a issue. Right now it works because `thrift call time` is much less than `idle timeout`.

Reply:
What do you mean by max-idle time? If you are referring scavenging the under connection, then in this case , object won't be present in the queue to get scavenged. 

All the objects are inherently same (don't have any state properties associated), setting timestamp while getting or returning shouldn't matter much. Or am I missing something?

Reply:
Agree with Shubham as to why it works for now.

> setting timestamp while getting or returning shouldn't matter much.

Ok, if this will not be a part of scavenger thread, then doesn't make sense to set the last access time in `release` method? `borrow` is logically not the last access time according to me.

Reply:
yeah, got it. makes sense. for example, if in a query burst all objects are taken for 20 minutes (suppose) and the released. Scanvenger will think these objects are not getting used and scavenge. 

We can take this in a separate commit as subhuam mentioned.